Init Process for AWS for Fluent Bit on ECS

Galaoaoa opened the pull request:
#### Description 

Users can build the new image  `amazon/aws-for-fluent-bit-init:latest` and original image `amazon/aws-for-fluent-bit:latest` as the same time by using the command  `make release`

---

#### Changes from last PR

Code Style

* Change the original code style to use a consistent code style and naming style with the team
* Change code structure for modularization and code reuse

Unit Test

* Add code for unit test in `fluent_bit_init_process_test.go`

Dockerfile

* Merge two dockerfiles into `Dockerfile.init`

---

#### Details of comparison with `aws/aws-for-fluent-bit @ mainline`

* Change `Makefile`
  Change the `make release` command, add `docker build -t amazon/aws-for-fluent-bit-init:latest -f Dockerfile.init .` After executing the `make release` command, will also build the image `amazon/aws-for-fluent-bit-init:latest`

* Create `Dockerfile.init`
  This Dockerfile will build the new image `amazon/aws-for-fluent-bit-init:latest` which include init process and finally execute `fluent_bit_init_entrypoint.sh`

* Create `init` folder
  This folder is used to store files related to the init process (following three files)

* Create `fluent_bit_init_entrypoint.sh`
  This file is used to run init process binary and run `fluent-bit-invoker.sh`, which will execute export command to set ECS Metadata, and finally invoke Fluent Bit

* Create `fluent_bit_init_process.go`
  This file is the code which used to achieve the desired init process functions

* Create `fluent_bit_init_process_test.go`
  This file is the code of Unit Test for `fluent_bit_init_process.go`

---

#### Tests already performed

**Run image on ECS**

* Without any user specified environment variables. Same effect as using the original image

* Add a Dummy Input config file via setting environment variables. Main config file is changed as expected

* Add a Dummy Input config file, a Parser config file and a related Filter config file via setting environment variables. Main config file is changed as expected, the Fluent Bit Command is changed as expected

* Add  a Dummy Input config file, a Parser config file, a related Filter config file, a S3 Output config file, a stdout Output config file via setting environment variables. Main config file is changed as expected, the Fluent Bit Command is changed as expected. Can receive the log in the container and S3

**Run image locally**

* Without ECS Task Metadata. Can invoke Fluent Bit and print out a warning `[FluentBit Init Process] Unable to get ECS Metadata, ignore this warning if not running on ECS`

zhonghui12 commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+	res, err := httpClient.Get(ecsTaskMetadataEndpointV4 + "/task")
+	if err != nil {
+		logrus.Warnln(err)
+		logrus.Warnln("[FluentBit Init Process] Failed to get ECS Metadata via HTTP Get")
```
For same level, why don't we merge them like `logrus.Warnln("[FluentBit Init Process] Failed to get ECS Metadata via HTTP Get: %s", err)`

zhonghui12 replied:
and I think it should be error level. If we cannot get it, we should directly return right? The logic should be the same as the file I gave you..

Galaoaoa replied:
Yeah, for the warning message, will merge them.

For the second question, I remember that you said inject ECS Metadata should be an optional.. So when running on ECS, init process can get the endpoint, but if run this image locally, customers can also use it normally just can't use metadata as env vars..

zhonghui12 replied:
yeah but I don't think error level will exit the code immediately?

zhonghui12 replied:
because you will directly return here if you are not in ECS and cannot get the endpoint: https://github.com/aws/aws-for-fluent-bit/pull/379/files/8c9902211f1d542a7ba372724898dbb62aca94bd#diff-a24b896daae00708d98524f520e1d75af05aea6586bd9e70568e44197652681e. So in this line, if we cannot get metadata from the endpoint, we should be in an error right?

zhonghui12 commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+
+	response, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		logrus.Warnln(err)
```
+1 same as above

zhonghui12 commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+	err = json.Unmarshal(response, &metadata)
+	if err != nil {
+		logrus.Warnln(err)
+		logrus.Warnln("[FluentBit Init Process] Failed to unmarshal ECS metadata")
```
same as above

zhonghui12 commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+	arn, err := arn.Parse(metadata.ECS_TASK_ARN)
+	if err != nil {
+		logrus.Warnln(err)
+		logrus.Warnln("[FluentBit Init Process] Failed to parse ECS TaskARN")
```
same as above

zhonghui12 commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+}
+
+// create a file, when flag is true, the file will be closed automatically after creation
+func createFile(filePath string, AutoClose bool) *os.File {
```
when autoclose will be set to false?

Galaoaoa replied:
https://github.com/aws/aws-for-fluent-bit/pull/379/files#diff-a24b896daae00708d98524f520e1d75af05aea6586bd9e70568e44197652681eR273

zhonghui12 replied:
then we will not close it after downloading content?

zhonghui12 commented on fluent-bit-init-process_test.go:
```diff
@@ -0,0 +1,295 @@
+	os.Setenv("ECS_CONTAINER_METADATA_URI_V4", "http://169.254.170.2/v4/50379854-baeb-4b84-9010-dd3fe2df5f20")
+
+	// Test case 1: full metadata
+	mockResponse1 := (`{"Cluster":"ygloaTest","TaskARN":"arn:aws:ecs:us-west-2:546265451413:task/ygloaTest/4ca5a280e68947cd84a8357f0d008fb5","Family":"code_test_A","Revision":"35","DesiredStatus":"RUNNING","KnownStatus":"RUNNING","PullStartedAt":"2022-07-06T03:06:45.282195951Z","PullStoppedAt":"2022-07-06T03:06:47.089531338Z","AvailabilityZone":"us-west-2a","LaunchType":"EC2","Containers":[{"DockerI ...
```
we can change the name to be more generate or official rather than using our alias. Like `Cluster:ecs-cluster`

Galaoaoa replied:
Yea make sense!

zhonghui12 commented on fluent-bit-init-process_test.go:
```diff
@@ -0,0 +1,295 @@
+package main
+
```
for name, let' do all `-` or `_` to keep consistent

Galaoaoa replied:
Mm-hmm, it's because Go unit test file must end with `_test.go`

zhonghui12 replied:
I see. Then I would prefer all `_`

zhonghui12 reviewed (commented):
Much better than last PR. good work

Galaoaoa commented on Dockerfile.init:
```diff
@@ -0,0 +1,150 @@
+FROM public.ecr.aws/amazonlinux/amazonlinux:latest as builder
```
Yea make sense. I also plan to continue to improve this part and will try again

Galaoaoa commented on fluent-bit-init-entrypoint.sh:
```diff
@@ -0,0 +1,3 @@
+./fluent-bit-init-process
```
Do you think it's a good idea to put all init-related files and folders in one subfolder like `fluent-bit-init`?

zhonghui12 replied:
I think we can just create a folder called `init` and put all related files under it

Galaoaoa commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+
+// static paths
+const (
+	s3FileFolderPath           = "fluent-bit-init-s3-files"
```
Will modify this with above comment. Actually I will modify all paths of files and folders to make them more reasonable

Galaoaoa commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+}
+
+type ECSTaskMetadata struct {
+	AWS_REGION          string `json:"AWSRegion"`      // AWS region
```
Will delete unnecessary comments..

Galaoaoa commented on fluent-bit-init-process.go:
```diff
@@ -0,0 +1,402 @@
+// static paths
+const (
+	s3FileFolderPath           = "fluent-bit-init-s3-files"
+	mainConfigFilePath         = "fluent-bit-init.conf"
```
Yeah will modify

Galaoaoa commented:
> Much better than last PR. good work

Thank you for your comments on this PR!
Some details need to be confirmed by your reply again :)

zhonghui12 commented on init/fluent_bit_init_entrypoint.sh:
```diff
@@ -0,0 +1,3 @@
+./init/fluent_bit_init_process
+cat /init/fluent-bit-init.conf
+source /init/fluent-bit-invoker.sh
```
nit: `invoker` sounds weird to me. I prefer `/init/init.sh`

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,400 @@
+	s3FileFolderPath       = "/init/fluent-bit-init-s3-files/"
+	mainConfigFile         = "/init/fluent-bit-init.conf"
+	originalMainConfigFile = "/fluent-bit/etc/fluent-bit.conf"
+	invokerFile            = "/init/fluent-bit-invoker.sh"
```
nit: naming issue as above

Galaoaoa commented on init/fluent_bit_init_process.go:
```diff
@@ -132,23 +128,24 @@ func setECSTaskMetadata(metadata ECSTaskMetadata, filePath string) {
+// create a folder to store S3 config files user specified (download them from S3)
 func createS3ConfigFileFolder(folderPath string) {
-	os.Mkdir(folderPath, os.ModePerm)
+	os.Mkdir(folderPath, 0777)
```
All file permissions like this will be reconfirmed

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,400 @@
+}
+
+// create Fluent Bit command to use "-c" to specify the new main config file
+func createCommand(command *string, filePath string) {
```
Do we need this extra function? Why don't we directly create a command like `exec /fluent-bit/bin/fluent-bit -e /fluent-bit/firehose.so -e /fluent-bit/cloudwatch.so -e /fluent-bit/kinesis.so -c /init/fluent-bit-init.conf`?

zhonghui12 replied:
oh ok

Galaoaoa commented on Dockerfile.init:
```diff
@@ -0,0 +1,30 @@
+FROM public.ecr.aws/amazonlinux/amazonlinux:latest as fluent-bit-initer
```
Will modify all these names

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,400 @@
+		updateCommand(path)
+	} else {
+		// this is not a parser config file. @INCLUDE
+		writeInclude(path, mainConfigFile)
```
I think we lack one thing here. For built-in file, we should check if it exists in our container right? We cannot just @include it. It should under this folder https://github.com/aws/aws-for-fluent-bit/tree/mainline/ecs or it should have been imported into aws-for-fluent-bit container already

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,400 @@
+		envKey = string(env_kv[0])
+		envValue = string(env_kv[1])
+
+		matched_s3, _ := regexp.MatchString("aws_fluent_bit_s3_", envKey)
```
Based on discussion, let's change this to `aws_fluent_bit_init_s3` and `aws_fluent_bit_init_file`

Galaoaoa replied:
And both `aws_fluent_bit_init_s3` and `aws_fluent_bit_init_S3` will be accepted as config files from S3.
I think it would be more helpful here not to use case sensitivity, customers may not be concerned about the way **s3** or **S3** is written

Galaoaoa commented:
**Update**

* Change the prefix from `aws_fluent_bit_s3|file` to `aws_fluent_bit_init_s3|file`
* Modified some of the naming that was not good enough
* Modified file permissions

DrewZhang13 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	metadataReigon string = ""
+)
+
+// HTTPClient interface
```
useless comment, could remove

PettitWesley replied:
actually go lint will require a comment for structures... preferably more descriptive than this but this is the minimum needed to pass the linter

Galaoaoa replied:
For these interfaces, I think they are useless comment. And maybe we can discuss a standard for our team's code comments, although this is not a very important thing for this moment..

DrewZhang13 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	Get(url string) (*http.Response, error)
+}
+
+// S3Downloader interface
```
useless comment, could remove.

PettitWesley replied:
See my comment on  HTTPClient

DrewZhang13 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	fileFromS3 := createFile(s3FileFolder+s3FileName[len(s3FileName)-1], false)
+	defer fileFromS3.Close()
+
+	_, err := s3Downloader.Download(fileFromS3,
```
Do we need to consider retry is download from S3 fail for once?

PettitWesley replied:
Check the docs, I think the downloader may have some setting to auto-add retries.

Galaoaoa replied:
Already modified. Thanks!

DrewZhang13 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	createFile(invokeFile, true)
+
+	// get ECS Task Metadata and set the region for S3 client
+	httpClient := createHTTPClient()
+	metadata := getECSTaskMetadata(httpClient)
+	metadataReigon = reflect.ValueOf(metadata).Field(0).Interface().(string)
+
```
I like the format in `main()`, it's clean and in good shape. 
These three lines could be optimized to only one function like others in `main()` as well.

and move these three lines into that function outside of `main()`.

Galaoaoa replied:
Writing them out for unit testing, `metadataReigon` has removed a from the `main()`. 
But I think I still have to keep the other two..

zhonghui12 commented on Dockerfile.init:
```diff
@@ -0,0 +1,29 @@
+
+RUN curl -sL -o /bin/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
+RUN chmod +x /bin/gimme
+RUN yum upgrade -y && yum install -y tar gzip git make gcc
```
do we need all of these packages?

Galaoaoa replied:
After verification, `tar gzip git` are required. Already remove `make gcc`. Thanks for this reminder~

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+		envKey = string(env_kv[0])
+		envValue = string(env_kv[1])
+
+		matched_s3, _ := regexp.MatchString("aws_fluent_bit_init_s3", envKey)
```
you can just make it case-insensitive here: https://stackoverflow.com/questions/15326421/how-do-i-do-a-case-insensitive-regular-expression-in-go. as well as file/File maybe

Galaoaoa replied:
Already modified. Thank you for this advice

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+
+func processConfigFile(path string) {
+
+	content_b, err := ioutil.ReadFile(path)
```
what does `content_b` mean? Avoid to use meaningless name

PettitWesley replied:
I think its always better to do full names if needed: `contentBytes`

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	exists3Client bool = false
+
+	// global ecs metadata region
+	metadataReigon string = ""
```
`metadataRegion`

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+
+	// global s3 client and flag
+	s3Client      *s3.S3
+	exists3Client bool = false
```
nit: `existS3Client`

PettitWesley replied:
Actually, `s3ClientExists` or `s3ClientCreated` would be more idiomatic I think... this fits what I've seem from other go programmers, if you have a bool, you use works like has or is or something else, but you put it in the variable name in a way that fits how you'd say it out loud in english, so `hass3Client` but that sounds weird to me... I prefer `s3ClientCreated` since its readable is exactly what you mean- has the S3 client been created or not yet?

Galaoaoa replied:
Already modified this variable name. Thanks for advice

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+		createS3Client()
+	}
+
+	// e.g. "arn:aws:s3:::ygloa-bucket/s3_parser.conf"
```
-> e.g. "arn:aws:s3:::user-bucket/s3_parser.conf"
Don't include alias here

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	}
+}
+
+func processS3ConfigFiles(folderPath string) {
```
I don't think it is needed. In function getAllConfigFiles(), from line 166, you can directly do:
```
if s3 {
 download S3;
return filePath;
}

processConfigFile(filePath); 
```

This function is very similar to `processConfigFile`

zhonghui12 replied:
Then you don't need to use one extra function to create a s3Folder, you can just save all S3 files in one path which shares the same prefix. I remember the command will automatically create one path or find that path and create the file.

Galaoaoa replied:
For #1 comment, thanks for this advice. Reuse `processConfigFile()` and delete `processS3ConfigFiles()` makes my code more concise

For #2 comment, after trying, `os.Create` cannot create a file in a non-existent folder, which means I need create this folder first..

zhonghui12 replied:
for #2, I think you can have a function for it.
```
// create a file from a specific path
func create(p string) (*os.File, error) {
	if err := os.MkdirAll(filepath.Dir(p), 0600); err != nil {
		return nil, err
	}
	return os.Create(p)
}
```
I believe  above should work? it will check if it needs to create a directory and then create the file

Galaoaoa replied:
I think it should work, will try it later, thanks

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	// add @INCLUDE in main config file to include original main config file
+	writeInclude(originalMainConfigFile, mainConfigFile)
+
+	// create Fluent Bit command to use "-R" to specify new main config file
```
nit: should be `-c`?

zhonghui12 commented on init/fluent_bit_init_process_test.go:
```diff
@@ -0,0 +1,298 @@
+	os.Setenv("ECS_CONTAINER_METADATA_URI_V4", "http://169.254.170.2/v4/50379854-baeb-4b84-9010-dd3fe2df5f20")
+
+	// Test case 1: full metadata
+	mockResponse1 := (`{"Cluster":"ecs-cluster","TaskARN":"arn:aws:ecs:us-west-2:546265451413:task/ecs-cluster/4ca5a280e68947cd84a8357f0d008fb5","Family":"code_test_A","Revision":"35","DesiredStatus":"RUNNING","KnownStatus":"RUNNING","PullStartedAt":"2022-07-06T03:06:45.282195951Z","PullStoppedAt":"2022-07-06T03:06:47.089531338Z","AvailabilityZone":"us-west-2a","LaunchType":"EC2","Containers":[{"Doc ...
```
I think this line is too long.. you can skip some useless parts.

Galaoaoa replied:
Actually I used ECS Exec specifically to get this real response for testing.. 

I want to keep this because real test cases are more likely to find potential bugs, and this will be useful for further testing. If we change the Metadate structure future, we can still use this test case to test it

zhonghui12 replied:
Then I prefer you put the two long responses in one place like in the very beginning as `const`. It is too long to see here

Galaoaoa replied:
Yea already modified~

zhonghui12 commented on init/fluent_bit_init_process_test.go:
```diff
@@ -0,0 +1,298 @@
+	assert.Equal(t, actualOutput1, expectedOutput1)
+
+	// Test case 2: Task launch type will be an empty string if container agent is under version 1.45.0
+	mockResponse2 := (`{"Cluster":"ecs-cluster","TaskARN":"arn:aws:ecs:us-west-2:546265451413:task/ecs-cluster/4ca5a280e68947cd84a8357f0d008fb5","Family":"code_test_A","Revision":"35","DesiredStatus":"RUNNING","KnownStatus":"RUNNING","PullStartedAt":"2022-07-06T03:06:45.282195951Z","PullStoppedAt":"2022-07-06T03:06:47.089531338Z","AvailabilityZone":"us-west-2a","Containers":[{"DockerId":"8c928beb55e ...
```
see above

zhonghui12 commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+
+	// process all config files in S3 config file folder
+	// add @INCLUDE in main config file and change the fluent bit command if it's a Parser config file
+	processS3ConfigFiles(s3FileFolderPath)
```
I don't think line 385 and 394 is needed. See my comment in line 285

zhonghui12 replied:
https://github.com/aws/aws-for-fluent-bit/pull/379#discussion_r923941368

PettitWesley commented on Dockerfile.init:
```diff
@@ -0,0 +1,30 @@
+RUN go mod init fluent_bit_init_process \
+    && go mod tidy \
+    && go build fluent_bit_init_process.go
+
```
you should not run `go mod` commands inside of a build script IMO/AFAIK. The go.mod and go.sum files are generally committed to the source control in most projects I have seen and this helps ensure all build use the same versions of dependencies.

Galaoaoa replied:
Yea, I changed it to copy `go.mod` and `go.sum` from our repo instead of creating a new one

PettitWesley commented on Makefile:
```diff
@@ -17,6 +17,7 @@ all: release
 release:
 	docker build --no-cache -t aws-fluent-bit-plugins:latest -f Dockerfile.plugins .
 	docker build -t amazon/aws-for-fluent-bit:latest -f Dockerfile .
+	docker build -t amazon/aws-for-fluent-bit-init:latest -f Dockerfile.init .
```
I thought/think the image should be `amazon/aws-for-fluent-bit:latest-init` or `amazon/aws-for-fluent-bit:init-latest`. 

Building the image with this tag: `amazon/aws-for-fluent-bit-init` implies that we have a Docker Hub repo with that name. But we do not. I beleive we are planning to push these images to the same repos as we use for the non-init image right? And thus only the tag is different. So the same hsould be true in our build script IMO.

PettitWesley replied:
So unfortunately creating a new Docker Hub repo is a whole big process... also while I understand the tag becomes very long, I also think its better if all AWS for Fluent Bit images are in one repo if possible. 

A good example of long tags is fluentd repo which has tons of tags: https://hub.docker.com/r/fluent/fluentd/tags

Galaoaoa replied:
Yea I already discussed this with @zhonghui12 .
Understand to create a new Docker Hub repo will be a big process..

We finally decided to use same repo and these tags:
`amazon/aws-for-fluent-bit:init-latest`
` amazon/aws-for-fluent-bit:init-2.26.0`
`amazon/aws-for-fluent-bit:init-amd64-2.26.0`

PettitWesley commented on init/fluent_bit_init_entrypoint.sh:
```diff
@@ -0,0 +1,3 @@
+./init/fluent_bit_init_process
+cat /init/fluent-bit-init.conf
```
Why cat the file in entrypoint? This means it will be printed to the users log... which isn't bad but its also not really needed... and the user can get Fluent Bit to output all registered plugins if htey run it with debug logging...

Galaoaoa replied:
This came from a discussion with Bingfeng, he asked me how customer can quickly know which config files they are using.. Maybe using `cat` here is not suitable. Will delete

PettitWesley commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+)
+
+var (
+	// default Fluent Bit command
+	fluentBitCommand = "exec /fluent-bit/bin/fluent-bit -e /fluent-bit/firehose.so -e /fluent-bit/cloudwatch.so -e /fluent-bit/kinesis.so"
```
nit: I think "base command" is a better way to describe this since you start with it and add to it

PettitWesley commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	ECS_TASK_ID         string `json:"TaskID"`
+	ECS_FAMILY          string `json:"Family"`
+	ECS_LAUNCH_TYPE     string `json:"LaunchType"`     // Task launch type will be an empty string if container agent is under version 1.45.0
+	ECS_REVISION        string `json:"Revision"`       // Revision number
+	ECS_TASK_DEFINITION string `json:"TaskDefinition"` // TaskDefinition = "family:revision"
```
Hmm why do we give folks one with just revision, and one with "family:revision", but not just family?? Can we add that please. Since we do not have service name available, family (without the revision, since that changes a lot) is commonly used to replace service name.

Galaoaoa replied:
We already have [:)](https://github.com/aws/aws-for-fluent-bit/pull/379/files#diff-e6468b32643a6ec3230488b9625cb4932b4e8ba605ac967778ec52bd790a00c7R57)

PettitWesley commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	ECS_TASK_DEFINITION string `json:"TaskDefinition"` // TaskDefinition = "family:revision"
+}
+
+func createHTTPClient() HTTPClient {
+	client := &http.Client{}
+	return client
+}
```
since this function is only one line, it does not really provide value for code organization I think... you could replace the one line call to this function with the one line to create the client.

PettitWesley commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+
+// get ECS Task Metadata via endpoint V4
+func getECSTaskMetadata(httpClient HTTPClient) ECSTaskMetadata {
+
```
nit: newline at beginning of function

PettitWesley commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+
+	response, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		logrus.Fatalf("[FluentBit Init Process] Failed to read HTTP response: %s\n", err)
```
nit: might be more clear to users if its: Failed to read ECS metadata  HTTP response.

PettitWesley commented on init/fluent_bit_init_process.go:
```diff
@@ -0,0 +1,401 @@
+	*command = *command + " -c " + filePath
+}
+
+// create a folder to store S3 config files user specified (download them from S3)
+func createS3ConfigFileFolder(folderPath string) {
+	os.Mkdir(folderPath, 0700)
```
nit: in code, it is preferable IMO to use "directory" instead of folder. Directory is the old unix term, folder came around once we had GUIs.

Galaoaoa replied:
Yea, already modified. Will pay attention when writing code in the future

Galaoaoa commented on Dockerfile.init:
```diff
@@ -0,0 +1,29 @@
+RUN chmod +x /bin/gimme
+RUN yum upgrade -y && yum install -y tar gzip git make gcc
+ENV HOME /home
+RUN /bin/gimme 1.17.9
+ENV PATH ${PATH}:/home/.gimme/versions/go1.17.9.linux.arm64/bin:/home/.gimme/versions/go1.17.9.linux.amd64/bin
```
I've tried switching to the stable version, but this prevents me from setting the `PATH` for the go command in the Dockerfile..
Using 1.17.9 is because other Dockerfiles in our repo use this version

Galaoaoa commented:
**Update**

* Changed the image name from `amazon/aws-for-fluent-bit-init:latest` to `amazon/aws-for-fluent-bit:init-latest`

* Deleted `processS3ConfigFiles()`, changed to reuse `processConfigFile()`

* Modified the regex to implement case-insensitive of s3/S3, file/File

* Modified some variable names

* Removed two useless packages `make gcc` in the Dockerfile

Galaoaoa commented:
**Update**

* Set the environment variable `FLB_AWS_USER_AGENT=init` to get the `amazon/aws-for-fluent-bit:init-latest` usage

* Deleted the `createS3ConfigFileDirectory()` function

* Modified the corresponding unit test

zhonghui12 commented on init/fluent_bit_init_process_test.go:
```diff
@@ -0,0 +1,300 @@
+	defer os.RemoveAll(s3FileDirectoryPathTest)
+
+	s3FilePath := "ygloa/files/aaa.conf"
+	bucketName := "ygloaBucket"
```
please update the username here

Galaoaoa replied:
already removed Alias in the unit test, thank you!

zhonghui12 reviewed (commented):
Overall looks good to me. Good work! 

Also, please rebase all commits into 1 before merge
